# Inductive Conformal Prediction

In [90]:
from sklearn.datasets import load_diabetes
from sklearn.model_selection import train_test_split
import numpy as np
import math
import matplotlib.pyplot as plt
from sklearn.model_selection import cross_val_score

In [27]:
diabetes = load_diabetes()
X_train, X_test, y_train, y_test = train_test_split(diabetes.data, diabetes.target, random_state=1201)

### Part 3

In [28]:
from sklearn.linear_model import Lasso
lasso = Lasso().fit(X_train,y_train)
lasso.score(X_train,y_train)

0.36801497049625675

In [29]:
lasso.score(X_test,y_test)

0.3528762069100633

In [30]:
diabetes['feature_names']

['age', 'sex', 'bmi', 'bp', 's1', 's2', 's3', 's4', 's5', 's6']

In [31]:
diabetes['data'].shape

(442, 10)

+ Training R^2 for the Lasso model, lasso score: 0.36801497049625675
+ Test R^2 for the Lasso model, lasso score: 0.4736545705189427
+ There are 10 features
+ The feature names are: 'age', 'sex', 'bmi', 'bp', 's1', 's2', 's3', 's4', 's5', 's6'


### Part 4

In [32]:
X = np.genfromtxt("diabetes.data", delimiter="	", skip_header = 1,
usecols=np.arange(10),)

y = np.genfromtxt("diabetes.data", delimiter="	", skip_header = 1,
usecols=10, dtype='int')

### Part 5

In [33]:
X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=1201)

### Part 6

In [34]:
diabetes = np.genfromtxt('diabetes.data',  skip_header = 1,
usecols=np.arange(10),)
diabetes.shape

(442, 10)

In [35]:
X[:2]

array([[ 59.    ,   2.    ,  32.1   , 101.    , 157.    ,  93.2   ,
         38.    ,   4.    ,   4.8598,  87.    ],
       [ 48.    ,   1.    ,  21.6   ,  87.    , 183.    , 103.2   ,
         70.    ,   3.    ,   3.8918,  69.    ]])

In [36]:
lasso = Lasso().fit(X_train, y_train)
lasso.score(X_train, y_train)

0.518311821269729

In [37]:
lasso.score(X_test,y_test)

0.4736545705189427

+ There are 10 features for the diabetes.data file
+ The feature names are AGE	SEX	BMI	BP	S1	S2	S3	S4	S5	S6
+ The lasso training score is 0.518311821269729
+ The lasso teset score is 

The lasso score is the same on the tes set but different on the training set

In [38]:
print(X_train.shape)
print(X_test.shape)

(331, 10)
(111, 10)


### Part 7 
(Note to self, using two seperate scalars is a perfect example of data snooping)

In [39]:
from sklearn.preprocessing import StandardScaler
scaler = StandardScaler()
# fit training set
scaler.fit(X_train)
# transform the training set and test set
scaler.transform(X_train)
scaler.transform(X_test)

array([[ 1.48660309, -0.9048101 ,  0.23994961, ..., -0.02266782,
         0.55554925, -0.0755863 ],
       [ 0.80307313,  1.10520428, -0.52093878, ..., -0.79300119,
        -0.31573303, -0.51918676],
       [ 1.03091645,  1.10520428,  0.06091705, ...,  0.74766556,
         0.30132132,  0.81161461],
       ...,
       [ 1.56255086, -0.9048101 ,  2.36596127, ..., -0.02266782,
         0.00814376, -0.0755863 ],
       [ 0.11954317,  1.10520428, -0.61045506, ..., -0.79300119,
        -0.3602469 ,  0.19057397],
       [ 1.48660309, -0.9048101 ,  1.42604032, ...,  0.8478089 ,
         1.655157  ,  2.31985616]])

### Part 8

In [40]:
lasso = Lasso().fit(X_train, y_train)
lasso.score(X_train, y_train)

0.518311821269729

In [41]:
lasso.score(X_test,y_test)

0.4736545705189427

Current Lasso scores are identical to those in part 6 but different from part 3. This is perhaps because we are using a Standard Scalar so we would expect the data to be different from that of the normalised data in part 3

### Part 9

In [42]:
plt.plot(lasso00001.coef_, '^', label="Lasso alpha=0.0001")
plt.plot(lasso001.coef_, 'v', label="Lasso alpha=0.01")
plt.plot(lasso.coef_, 's', label="Lasso alpha=1")
plt.legend(ncol=2,loc=(0,1.05))
plt.ylim(-25,25)
plt.xlabel("Coefficient index")
plt.ylabel("Coefficient magnitude")

NameError: name 'lasso00001' is not defined

### Part 10
Compute best value for α

In [73]:
best_score = 0
for alpha in [0.0001, 0.001, 0.01, 0.1, 1, 10, 100]:
    # for each parameter of alpha,
    # train a a lasso
    lasso = Lasso(alpha=alpha, max_iter=100000).fit(X_train,y_train)
    # perform cross-validation
    scores = cross_val_score(lasso, X_train, y_train)
    # compute mean cross-validation accuracy
    score = np.mean(scores)
    # if we got a better score, store the score and parameters
    if score > best_score:
      best_score = score
      best_alpha = alpha
# rebuild a model on the full training set
lasso = Lasso(alpha=best_alpha, max_iter=100000).fit(X_train,y_train)
lasso.fit(X_train, y_train)
lasso_score = lasso.score(X_test, y_test)
print("Best CV score:", best_score)
print("Best parameter for alpha:", best_alpha)
print("Test set score with best parameter:", lasso_score)

Best CV score: 0.49008256025422375
Best parameter for alpha: 0.001
Test set score with best parameter: 0.4765818304860368


## Part 11

#### a)  

In [68]:
X = np.genfromtxt("diabetes.data", delimiter="	", skip_header = 1,
usecols=np.arange(10),)

y = np.genfromtxt("diabetes.data", delimiter="	", skip_header = 1,
usecols=10, dtype='int')

X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=1201)

In [69]:
# CS = calibration set
X_TSP, X_CS, y_TSP, y_CS = train_test_split(X_train, y_train, random_state=1201, test_size=99)

#### b)

In [153]:
scaler = StandardScaler()
scaler.fit(X_TSP)
scaler.transform(X_TSP)
scaler.transform(X_CS)
scaler.transform(X_test)

array([[ 1.46756957, -0.89356008,  0.17269573, ..., -0.09941834,
         0.50925478, -0.16429168],
       [ 0.78220046,  1.11911893, -0.59131094, ..., -0.87628112,
        -0.35443511, -0.60853026],
       [ 1.01065683,  1.11911893, -0.00707055, ...,  0.67744443,
         0.2572422 ,  0.72418548],
       ...,
       [ 1.54372169, -0.89356008,  2.30742025, ..., -0.09941834,
        -0.0333806 , -0.16429168],
       [ 0.09683136,  1.11911893, -0.68119408, ..., -0.87628112,
        -0.39856108,  0.10225147],
       [ 1.46756957, -0.89356008,  1.3636473 , ...,  0.77843659,
         1.59928049,  2.23459666]])

#### c)
nonconformity measure α = |y − ŷ|

test error rate := total number of failures / test_set. Where failure is the prediction interval failing to cover the true label for that test sample.

In [82]:
y_CS

array([275, 145, 158, 166, 101,  49, 161,  49, 248, 242, 288,  71, 168,
       220,  71, 275,  65, 116, 182,  77,  84, 129, 214, 248, 142, 235,
       162, 151, 280, 150,  51,  72, 219, 110,  66, 185,  48,  52, 202,
       125, 178, 306,  72,  59, 163, 144, 303,  55,  94, 102, 302,  88,
        91,  84, 148,  63, 141,  45,  93, 182, 118, 173,  72,  94,  66,
       131,  97,  65, 181,  77,  40, 236,  55, 135, 202, 115, 101, 140,
       245, 215,  71,  80, 131,  53, 136,  75, 122, 214, 268,  43,  97,
       170, 102,  79, 181,  25, 237,  78, 283])

In [117]:
#Non conformity scoring
def non_conformity(y, y_hat):
    score = y - y_hat
    score = math.fabs(score) #score = abs(score) for integer
    return score

In [220]:
#RAW PREDICTION STEP
#compute nonconformity score for each in calbration set (CS) and test set (test)
calibration_predictions = lasso.predict(X_CS)
test_predictions = lasso.predict(X_test)

calibration_set = []
for label in range(len(calibration_predictions)):
    calibration_set.append(non_conformity(y_CS[label], calibration_predictions[label]))
    
test_scores = []
for label in range(len(test_predictions)):
    test_scores.append(non_conformity(test_predictions[label], y_test[label]))

In [223]:
#rank non conformity score from test sample and compute p_value
def p_values(calibration_set_scores, samples_score):
    p_value = 0
    score = len(calibration_set_scores)+1
    
    for i in range(len(calibration_set_scores)):
        if samples_score < calibration_set_scores[i]:
            score -= 1
            
    p_value = score/(len(calibration_set_scores)+1)
    return p_value
    
    

In [241]:
#CALIBRATION STEP
#compute p-values for each in test set
p_values_list = []


for score in range(len(test_predictions)):
    p_values_list.append(p_values(calibration_set, test_predictions[score]))


In [243]:
def compute_error_rate(epsilon = 0.95):
    error = 0
    
    for p_value in range(len(p_values_list)):
        if p_values_list[p_value] < epsilon:
            error += 1
    error_rate = error / len(test_predictions)
    return error_rate

In [ ]:
#OUTPUTS
#(the predicion intervals paired with each TSP sample)

In [245]:
#TEST ERROR RATE
print('error rate at 5% threshold: ',compute_error_rate(0.95))
print('error rate at 20% threshold: ',compute_error_rate(0.80))

error rate at 5% threshold:  0.18018018018018017
error rate at 20% threshold:  0.018018018018018018
